File: apps/cargo-tracker/internal/shipper/application/register_shipper_service.go
package application

import (
	"context"
	"errors"

	"github.com/k2works/case-study-cargo-tracker/apps/cargo-tracker/internal/shipper/domain"
)

// ErrEmailAlreadyRegistered は同一メールアドレスが既に登録済みの場合に返される。
var ErrEmailAlreadyRegistered = errors.New("email already registered")

// ErrCorporateContractRequired は法人荷主に契約情報が欠けている場合に返される。
var ErrCorporateContractRequired = errors.New("corporate shipper requires contract number and discount rate")

// RegisterShipperCommand は荷主登録のコマンド。
type RegisterShipperCommand struct {
	Name           string
	Email          string
	ShipperType    string
	Phone          *string
	Address        *string
	ContractNumber *string
	DiscountRate   *float64
}

// RegisterShipperService は荷主登録ユースケースを実行するコマンドサービス。
type RegisterShipperService struct {
	repo  ShipperRepository
	idGen IDGenerator
}

// NewRegisterShipperService は RegisterShipperService を生成する。
func NewRegisterShipperService(repo ShipperRepository, idGen IDGenerator) *RegisterShipperService {
	return &RegisterShipperService{repo: repo, idGen: idGen}
}

// Register は荷主を新規登録し、発行された ShipperId を返す。
func (s *RegisterShipperService) Register(ctx context.Context, cmd RegisterShipperCommand) (domain.ShipperId, error) {
	email, err := domain.NewEmail(cmd.Email)
	if err != nil {
		return domain.ShipperId{}, err
	}

	exists, err := s.repo.ExistsByEmail(ctx, email)
	if err != nil {
		return domain.ShipperId{}, err
	}
	if exists {
		return domain.ShipperId{}, ErrEmailAlreadyRegistered
	}

	name, err := domain.NewShipperName(cmd.Name)
	if err != nil {
		return domain.ShipperId{}, err
	}

	shipperType, err := domain.ParseShipperType(cmd.ShipperType)
	if err != nil {
		return domain.ShipperId{}, err
	}

	rawUUID := s.idGen.Generate()
	id, err := domain.NewShipperId(rawUUID)
	if err != nil {
		return domain.ShipperId{}, err
	}
	code := domain.GenerateShipperCode(rawUUID)

	var shipper *domain.Shipper
	if shipperType == domain.ShipperTypeCorporate {
		shipper, err = s.buildCorporate(id, code, name, email, cmd)
	} else {
		shipper, err = domain.RegisterIndividualShipper(id, code, name, email)
	}
	if err != nil {
		return domain.ShipperId{}, err
	}

	s.applyOptional(shipper, cmd)

	if err := s.repo.Save(ctx, shipper); err != nil {
		return domain.ShipperId{}, err
	}
	return id, nil
}

func (s *RegisterShipperService) buildCorporate(
	id domain.ShipperId, code domain.ShipperCode, name domain.ShipperName, email domain.Email, cmd RegisterShipperCommand,
) (*domain.Shipper, error) {
	if cmd.ContractNumber == nil || cmd.DiscountRate == nil {
		return nil, ErrCorporateContractRequired
	}
	contract, err := domain.NewContractNumber(*cmd.ContractNumber)
	if err != nil {
		return nil, err
	}
	rate, err := domain.NewDiscountRate(*cmd.DiscountRate)
	if err != nil {
		return nil, err
	}
	return domain.RegisterCorporateShipper(id, code, name, email, contract, rate)
}

func (s *RegisterShipperService) applyOptional(shipper *domain.Shipper, cmd RegisterShipperCommand) {
	if cmd.Phone != nil {
		shipper.SetPhone(domain.NewPhone(*cmd.Phone))
	}
	if cmd.Address != nil {
		if addr, err := domain.NewAddress(*cmd.Address); err == nil {
			shipper.SetAddress(addr)
		}
	}
}

File: apps/cargo-tracker/src/contexts/shipper/application/commandservices/register-shipper.service.ts
import { Shipper } from '../../domain/model/shipper.js';
import { ShipperType } from '../../domain/model/value-objects.js';
import {
  EmailAlreadyRegisteredError,
  type ShipperRepository,
} from '../../domain/repository/shipper-repository.js';

export interface RegisterShipperCommand {
  shipperType: ShipperType;
  name: string;
  email: string;
  phone?: string;
  address?: string;
  contractNumber?: string;
  discountRate?: number;
}

/**
 * 荷主登録ユースケース（US02/US03）。
 * Email 重複チェックと ShipperCode 自動生成（集約側）を行う。
 */
export class RegisterShipperService {
  constructor(private readonly shippers: ShipperRepository) {}

  async register(command: RegisterShipperCommand): Promise<number> {
    const existing = await this.shippers.findByEmail(command.email.trim().toLowerCase());
    if (existing !== null) {
      throw new EmailAlreadyRegisteredError(command.email);
    }

    const shipper = this.buildShipper(command);
    return this.shippers.save(shipper);
  }

  private buildShipper(command: RegisterShipperCommand): Shipper {
    if (command.shipperType === ShipperType.CORPORATE) {
      if (!command.contractNumber) {
        throw new Error('法人荷主には契約番号が必要です');
      }
      return Shipper.registerCorporate({
        name: command.name,
        email: command.email,
        phone: command.phone,
        address: command.address,
        contractNumber: command.contractNumber,
        discountRate: command.discountRate ?? 0,
      });
    }
    return Shipper.registerIndividual({
      name: command.name,
      email: command.email,
      phone: command.phone,
      address: command.address,
    });
  }
}
